File: cyanBat/src/main/java/at/smiech/cyanbat/activities/CyanBatGameActivity.java
package at.smiech.cyanbat.activities;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.os.Vibrator;
import androidx.annotation.Nullable;
import android.util.Log;
import android.view.Window;
import android.view.WindowManager;

import at.grueneis.game.framework.Graphics;
import at.grueneis.game.framework.Graphics.PixmapFormat;
import at.grueneis.game.framework.Music;
import at.grueneis.game.framework.Pixmap;
import at.grueneis.game.framework.Screen;
import at.grueneis.game.framework.Sound;
import at.grueneis.game.framework.code.AndroidGameActivity;
import at.smiech.cyanbat.screens.GameScreen;
import at.smiech.cyanbat.util.MusicPlayer;

public class CyanBatGameActivity extends AndroidGameActivity {

    public static final String TAG = "CyanBatGameActivity";
    public static final boolean DEBUG = false;
    public static final boolean SHOOTING_ENABLED = false;

    public static Pixmap bat;
    public static Pixmap death;
    public static Pixmap menuButtons;
    public static Pixmap mainMenu;
    public static Pixmap gameOver;
    public static Pixmap background;
    public static Pixmap[] topObstacles = new Pixmap[2];
    public static Pixmap[] bottomObstacles = new Pixmap[2];
    public static Music gameTrack;
    public static Sound deathSound;
    public static Music gameOverMusic;
    public static Pixmap enemies;
    public static Vibrator vib;
    public static MusicPlayer musicPlayer;
    public static Pixmap explosion;
    public static Pixmap shot;
    public static CyanBatGameActivity currentActivity = null;

    public Screen getStartScreen() {
        if (DEBUG)
            Log.d(TAG, "getStartScreen");
        currentActivity = this;
        initAssets();
        return new GameScreen(this);
    }

    private void initAssets() {
        if (DEBUG)
            Log.d(TAG, "initAssets");

        Graphics g = getGraphics();
        // Loading image assets
        bat = g.newPixmap("cyanBat.png", PixmapFormat.ARGB4444);
        gameOver = g.newPixmap("gameover.png", PixmapFormat.ARGB4444);
        background = g.newPixmap("background.jpg", PixmapFormat.ARGB4444);
        topObstacles[0] = g
                .newPixmap("topObstacle1.png", PixmapFormat.ARGB4444);
        topObstacles[1] = g
                .newPixmap("topObstacle2.png", PixmapFormat.ARGB4444);
        bottomObstacles[0] = g.newPixmap("bottomObstacle1.png",
                PixmapFormat.ARGB4444);
        bottomObstacles[1] = g.newPixmap("bottomObstacle2.png",
                PixmapFormat.ARGB4444);
        death = g.newPixmap("death.png", PixmapFormat.ARGB4444);
        enemies = g.newPixmap("enemies.png", PixmapFormat.ARGB4444);
        explosion = g.newPixmap("explosion.png", PixmapFormat.ARGB4444);
        shot = g.newPixmap("shot.png", PixmapFormat.ARGB4444);

        musicPlayer = new MusicPlayer();
        gameTrack = getAudio().newMusic("game_theme.mp3");
        gameOverMusic = getAudio().newMusic("game_over.mp3");
        deathSound = getAudio().newSound("deathSound.mp3");

        vib = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState, @Nullable PersistableBundle persistentState) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Window window = getWindow();
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.setNavigationBarColor(Color.BLACK);
        }

        super.onCreate(savedInstanceState, persistentState);
    }
}

File: cyanBat/src/main/java/at/smiech/cyanbat/activities/CyanBatGameActivity.kt
package at.smiech.cyanbat.activities

import android.content.Context
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.PersistableBundle
import android.os.Vibrator
import android.os.VibratorManager
import android.util.Log
import android.view.WindowManager
import at.grueneis.game.framework.Graphics.PixmapFormat
import at.grueneis.game.framework.Music
import at.grueneis.game.framework.Pixmap
import at.grueneis.game.framework.Screen
import at.grueneis.game.framework.Sound
import at.grueneis.game.framework.impl.AndroidGameActivity
import at.smiech.cyanbat.screen.GameScreen
import at.smiech.cyanbat.util.MusicPlayer

class CyanBatGameActivity : AndroidGameActivity() {

    override val startScreen: Screen
        get() {
            if (DEBUG)
                Log.d(TAG, "getStartScreen")
            initAssets()
            return GameScreen(this)
        }

    private fun initAssets() {
        if (DEBUG)
            Log.d(TAG, "initAssets")

        // Loading image assets
        graphics?.let { g ->
            bat = g.newPixmap("cyanBat.png", PixmapFormat.ARGB8888)
            gameOver = g.newPixmap("gameover.png", PixmapFormat.ARGB8888)
            background = g.newPixmap("background.jpg", PixmapFormat.ARGB8888)
            topObstacles[0] = g
                .newPixmap("topObstacle1.png", PixmapFormat.ARGB8888)
            topObstacles[1] = g
                .newPixmap("topObstacle2.png", PixmapFormat.ARGB8888)
            bottomObstacles[0] = g.newPixmap("bottomObstacle1.png",
                PixmapFormat.ARGB8888)
            bottomObstacles[1] = g.newPixmap("bottomObstacle2.png",
                PixmapFormat.ARGB8888)
            death = g.newPixmap("death.png", PixmapFormat.ARGB8888)
            enemies = g.newPixmap("enemies.png", PixmapFormat.ARGB8888)
            explosion = g.newPixmap("explosion.png", PixmapFormat.ARGB8888)
            shot = g.newPixmap("shot.png", PixmapFormat.ARGB8888)
        }

        musicPlayer = MusicPlayer()
        musicEnabled = sharedPrefs?.getBoolean("music_enabled", true) ?: true
        musicPlayer.setEnabled(musicEnabled)

        //music
        audio?.let {
            gameTrack = it.newMusic("game_theme.mp3")
            gameOverMusic = it.newMusic("game_over.mp3")
            deathSound = it.newSound("deathSound.mp3")
        }

        //sound
        soundsEnabled = sharedPrefs?.getBoolean("sounds_enabled", true) ?: true

        vib = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vibratorManager =
                getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
            vibratorManager.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            getSystemService(VIBRATOR_SERVICE) as Vibrator
        }


    }

    override fun onCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        window.navigationBarColor = Color.BLACK

        super.onCreate(savedInstanceState, persistentState)
    }


    override fun onResume() {
        // checking pref changes
        musicEnabled = this.sharedPrefs?.getBoolean("music_enabled", true) == true
        musicPlayer.setEnabled(musicEnabled)
        soundsEnabled = this.sharedPrefs?.getBoolean("sounds_enabled", true) == true
        super.onResume()
    }

    companion object {
        const val TAG = "CyanBatGameActivity"
        const val DEBUG = false
        const val SHOOTING_ENABLED = false

        lateinit var bat: Pixmap
        lateinit var death: Pixmap
        lateinit var gameOver: Pixmap
        lateinit var background: Pixmap
        var topObstacles = arrayOfNulls<Pixmap>(2)
        var bottomObstacles = arrayOfNulls<Pixmap>(2)
        lateinit var gameTrack: Music
        lateinit var deathSound: Sound
        lateinit var gameOverMusic: Music
        lateinit var enemies: Pixmap
        lateinit var vib: Vibrator
        lateinit var musicPlayer: MusicPlayer
        lateinit var explosion: Pixmap
        lateinit var shot: Pixmap
        var musicEnabled = true
        var soundsEnabled = true
    }
}
